src/connector.c

```diff
@@ -0,0 +1,319 @@
+                                         const char *host, const char *port, tlsuv_connect_cb cb, void *ctx);
+static void proxy_cancel(tlsuv_connector_req req);
+
+// prevent freeing default connector
+static void default_connector_free(void *self){
+    (void)self;
+}
+
+static tlsuv_connector_t default_connector = {
+        .connect = default_connect,
+        .cancel = default_cancel,
+        .free = default_connector_free,
+};
+static const tlsuv_connector_t *global_connector = &default_connector;
+
+struct tlsuv_proxy_connector_s {
+    tlsuv_connect connect;
+    void (*cancel)(tlsuv_connector_req);
+    void (*free)(tlsuv_connector_t *self);
+    
+    tlsuv_proxy_t type;
+    char *host;
+    char *port;
+};
+
+void tlsuv_set_global_connector(const tlsuv_connector_t *c) {
+    global_connector = c ? c : &default_connector;
+}
+
+const tlsuv_connector_t* tlsuv_global_connector() {
+    return global_connector;
+}
+
+static void on_close(uv_handle_t *h) {
+    struct conn_req_s *req = h->data;
+    free(req);
+}
+
+static void def_connect_work(uv_work_t *w) {
+    struct conn_req_s *r = container_of(w, struct conn_req_s, connect);
+
+    struct pollfd fds[] = {
+            {
+                    .fd = r->sock,
+                    .events = POLLOUT,
+                    .revents = 0,
+            }
+    };
+#if _WIN32
+    WSAPoll(fds, 1, -1);
+#else
+    poll(fds, 1, -1);
+#endif
+
+    if (fds[0].revents & POLLHUP) {
+        socklen_t sel = sizeof(r->error);
+        getsockopt(r->sock, SOL_SOCKET, SO_ERROR, &r->error, &sel);
+#if _WIN32
+        switch (r->error) {
+            case WSAECONNREFUSED: r->error = UV_ECONNREFUSED; break;
+            case WSAECONNRESET: r->error = UV_ECONNRESET; break;
+            case WSAECONNABORTED: r->error = UV_ECONNABORTED; break;
+            default:
+                r->error = -r->error;
+        }
+#else
+        r->error = -r->error;
+#endif
+    }
+}
+
+static void def_connect_done(uv_work_t *w, int status) {
+    struct conn_req_s *r = container_of(w, struct conn_req_s, connect);
+    if (r->cancel) {
+        r->error = UV_ECANCELED;
+    } else if (status != 0) {
+        r->error = status;
+    }
+
+    if (r->error != 0) {
+        closesocket(r->sock);
+        r->sock = (uv_os_sock_t)-1;
+    }
+
+    r->cb(r->sock, r->error, r->ctx);
+    free(r);
+}
+
+static void on_resolve(uv_getaddrinfo_t *r, int status, struct addrinfo *addr) {
+    struct conn_req_s *cr = container_of(r, struct conn_req_s, connect);
+
+    if (status != 0) {
+        cr->cb(cr->sock, status, cr->ctx);
+        free(r);
+    } else {
+        cr->sock = tlsuv_socket(addr, 0);
+        int rc = connect(cr->sock, addr->ai_addr, addr->ai_addrlen);
+        int e = get_error();
+        if (rc == 0 ||
+#if _WIN32
+            e == WSAEWOULDBLOCK
+#else
+            e == EINPROGRESS
+#endif
+                ) {
+            uv_queue_work(r->loop, &cr->connect, def_connect_work, def_connect_done);
```

Comment:
Could the same be achieved with uv_poll_t without using a worker thread?

Reply:
you're right, we should use uv_poll to prevent worker pool blocking